### DISCUSSION TITLE
Styling and theming in React Flow

### DISCUSSION THREAD

**Discussion by chrtze** [xyflow/xyflow#2306] [General]
We are planning on updating the css handling of React Flow. Right now, React Flow includes two stylesheets, one with the base styles that are mandatory for React Flow to work properly, one with very minimal theme styles that can lead as a starting point for your own styling.

Our ideas range from keeping it as minimal as possible (basically including just the base css so that everything works) to including a theming library like [stitches](https://stitches.dev/) which would make it possible to write your own theme and pass it as a prop.

We would be interested in how you are styling your React Flow apps at the moment and what could be improved to make it more convenient. Please share your thoughts and if you like, answer one of these questions:

- Are you using the [nocss entrypoint](https://reactflow.dev/docs/getting-started/installation/#usage)?
- How are you customizing the default styling?
- Is it convenient for you to style via class names?
- Do you have any suggestion / best practice example or another library where you liked the way they are handling the styles?
- How are you styling your custom nodes?
- Would it be helpful if React Flow would provide more default styles?

Thank you!

**Comment by moklick** [comment]
we want to keep it simple but also flexible that's why we probably go into this direction:

```js
// 1. standard bundle (default styling)

import { ReactFlow, Controls } from '@react-flow/bundle';

import '@react-flow/bundle/dist/style.css';

// 2. base bundle (only very basic styling)

import { ReactFlow, Controls } from '@react-flow/bundle';

import '@react-flow/bundle/dist/base.css';

// 3. standard separate package imports (default styling)

import { ReactFlow } from '@react-flow/core';
import { Controls } from '@react-flow/controls';

import '@react-flow/core/dist/style.css';
import '@react-flow/controls/dist/style.css';

// 4. separate package imports base only (only very basic styling)

import { ReactFlow } from '@react-flow/core';
import { Controls } from '@react-flow/controls';

import '@react-flow/core/dist/base.css';
import '@react-flow/controls/dist/style.css';
```

This would mean that you have to load the CSS yourself and we don't inject it magically anymore. The advantages are that it's easier to integrate into your build process, easier to overwrite and no magic anymore.

**Reply by himself65** [reply]
I would like the 3

**Reply by moklick** [reply]
all of these ways would be possible with the new version :)

**Reply by himself65** [reply]
> all of these ways would be possible with the new version :) 

thats perfect

**Comment by HazelDaniel** [comment]
so, what is the current update on this?.
custom styling has been hell, especially with the box model and positioning.
can you at least expose some css variables to work with?

**Reply by moklick** [reply]
You can find all information about theming here https://reactflow.dev/learn/customization/theming . We expose css variables for styling since v12.

**Comment by HazelDaniel** [comment]
It's still not enough for some use cases.
For example, I would like to add things like css transforms to the nodes myself but I noticed it's being used to automatically update the positions of the nodes on drag.

Unless there's a workaround for this (to which I'd appreciate a pointer) , I'd suggest exposing variables to indicate their current coordinates . E.g:
--curr-y
--curr-x

And have them local to each nodes so that we can apply other transforms outside of just the translate transforms.

**Reply by moklick** [reply]
That's not something we will add to the library. What are you trying to do?
